src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -6982,12 +6982,7 @@ private void VisitUsingVariableDeclarationOperation(IUsingDeclarationOperation o
             StartVisitingStatement(operation);
 
             // a using statement introduces a 'logical' block after declaration, we synthesize one here in order to analyze it like a regular using 
-            BlockOperation logicalBlock = new BlockOperation(
-                operations: statements,
-                locals: ImmutableArray<ILocalSymbol>.Empty,
-                ((Operation)operation).OwningSemanticModel,
-                operation.Syntax,
-                isImplicit: true);
+            BlockOperation logicalBlock = BlockOperation.CreateTemporaryBlock(statements, ((Operation)operation).OwningSemanticModel!, operation.Syntax);
```

Comment:
This is required because of the uncommented verification code in Operation.SetParentOperation. It was always failing, but we had commented it out because of race conditions, and so missed that this was an issue when it was added.

Reply:
> This is required because of the uncommented verification code in Operation.SetParentOperation. It was always failing, but we had commented it out because of race conditions, and so missed that this was an issue when it was added.

Could you elaborate please what exactly was failing and why?

---
In reply to: [526329457](https://github.com/dotnet/roslyn/pull/49453#discussion_r526329457) [](ancestors = 526329457)

Reply:
> Could you elaborate please what exactly was failing and why?

The constructor of `BlockOperation` calls `SetParentOperation`. Because of the race condition, we had commented out the verification code that ensured that the result of `SetParentOperation` was the node we were expecting. These statements already have parent operations, and so adding that verification back in failed here, because `SetParentOperation` did not set their parent nodes to this new Block.

---
In reply to: [526917375](https://github.com/dotnet/roslyn/pull/49453#discussion_r526917375) [](ancestors = 526917375,526329457)

Reply:
> The constructor of BlockOperation calls SetParentOperation

You mean for children that we are passing?

---
In reply to: [527277309](https://github.com/dotnet/roslyn/pull/49453#discussion_r527277309) [](ancestors = 527277309,526917375,526329457)

Reply:
> You mean for children that we are passing?

Yep. These children already have parents, so the Interlocked.CompareExchange does not set it to the new block and verification fails.

---
In reply to: [527279503](https://github.com/dotnet/roslyn/pull/49453#discussion_r527279503) [](ancestors = 527279503,527277309,526917375,526329457)

Reply:
I'm adding additional asserts to the CreateTemporaryBlock constructor to verify the scenario.

---
In reply to: [527283599](https://github.com/dotnet/roslyn/pull/49453#discussion_r527283599) [](ancestors = 527283599,527279503,527277309,526917375,526329457)